### PR TITLE
WIP -- COLO Coin Redemption

### PR DESCRIPTION
**Pull request by eswarasai**
##### Description

Added a page to allow users to redeem the COLO coin

##### Checklist

- [ ] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Refers/Fixes

Fixes: #261

### CODE REVIEW

**Comment by eswarasai** [comment]
@owocki -- I've wrapped up the pending changes for this task. Please find below steps to test the changes :
1) Run migrations -- `./manage.py migrate dashboard`
2) Set below local settings values appropriately -- 
```
# WEB3
WEB3_HTTP_PROVIDER = 'TODO'

# COLO Coin
COLO_ACCOUNT_ADDRESS = 'TODO'
COLO_ACCOUNT_PRIVATE_KEY = 'TODO
```
3) Create COLO CoinRedemption via below management command with the arguments `count`, `network`, `token_name`, `contract_address`, `amount`: (number of tokens), `expiry_date`: (in seconds)
```
./manage.py create_coin_redemption_shortcodes 5 ropsten COLO 0x2941dEaad71ADb02B944bd38EBce2f1F4C9A62Dc 1 604800
```

Let me know if there are any modifications/suggestions/issues that are needed to be taken care of. Thanks.

File: app/dashboard/views.py [hunk 1]
```diff
@@ -776,3 +782,75 @@ def toolbox(request):
...
+            tx = contract.functions.transfer(address, coin.amount * 10**16).buildTransaction({
+                'nonce': w3.eth.getTransactionCount(settings.COLO_ACCOUNT_ADDRESS),
+                'gas': 40000,
+                'gasPrice': recommend_min_gas_price_to_confirm_in_time(5) * 10**9
+            })
+
+            signed = w3.eth.account.signTransaction(tx, settings.COLO_ACCOUNT_PRIVATE_KEY)
+            transaction_id = w3.eth.sendRawTransaction(signed.rawTransaction)
+
+            crr = CoinRedemptionRequest.objects.create(
```
Comment (mbeacom) [inline app/dashboard/views.py]: You can drop the variable assignment here and simply do: `CoinRedemptionRequest.objects.create(`

File: app/dashboard/views.py [hunk 2]
```diff
@@ -776,3 +782,75 @@ def toolbox(request):
...
+            contract = w3.eth.contract(coin.contract_address, abi=abi)
+
+            tx = contract.functions.transfer(address, coin.amount * 10**16).buildTransaction({
+                'nonce': w3.eth.getTransactionCount(settings.COLO_ACCOUNT_ADDRESS),
+                'gas': 40000,
+                'gasPrice': recommend_min_gas_price_to_confirm_in_time(5) * 10**9
+            })
+
+            signed = w3.eth.account.signTransaction(tx, settings.COLO_ACCOUNT_PRIVATE_KEY)
+            transaction_id = w3.eth.sendRawTransaction(signed.rawTransaction)
```
Comment (mbeacom) [inline app/dashboard/views.py]: You could `transaction_id = w3.eth.sendRawTransaction(signed.rawTransaction).hex()` and reference `transaction_id` in place of calling `transaction_id.hex()` throughout.

File: app/dashboard/views.py [hunk 3]
```diff
@@ -31,18 +32,22 @@
 
 from app.utils import ellipses, sync_profile
 from dashboard.helpers import normalizeURL, process_bounty_changes, process_bounty_details
-from dashboard.models import Bounty, BountySyncRequest, Interest, Profile, ProfileSerializer, Subscription, Tip
+from dashboard.models import Bounty, BountySyncRequest, Interest, Profile, ProfileSerializer, Subscription, Tip, CoinRedemption, CoinRedemptionRequest
```
Comment (mbeacom) [inline app/dashboard/views.py]: Once syntax errors are resolved throughout the additions, please run `isort app/dashboard/views.py` from your `web` repository root directory, so the imports reflect something like:

```python
from dashboard.models import (
    Bounty, BountySyncRequest, CoinRedemption, CoinRedemptionRequest, Interest, Profile, ProfileSerializer,
    Subscription, Tip,
)
```
and
`from web3 import HTTPProvider, Web3`

File: app/assets/yge/redeem_coin.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+};
+
+function redeemCoin() {
+    mixpanel.track("Redeem Coin Click", {});
+    metaMaskWarning();
+
+    var forwarding_address = $("forwarding_address").value.trim();
+
+    // Check for valid checksum address
+    if (!forwarding_address || !(/^(0x)?[0-9a-f]{40}$/.test(forwarding_address) || /^(0x)?[0-9A-F]{40}$/.test(forwarding_address))) {
```
Comment (mbeacom) [inline app/assets/yge/redeem_coin.js]: Maybe we could use `if (!forwarding_address || !web3.isAddress(forwarding_address)) {` here?

File: app/dashboard/models.py [hunk 5]
```diff
@@ -764,3 +764,21 @@ def normalize_tip_usernames(sender, instance, **kwargs):
...
+
+
+class CoinRedemption(SuperModel):
+  shortcode = models.CharField(max_length=255, default='')
+  url = models.URLField(null=True)
+  network = models.CharField(max_length=255, default='')
+  token_name = models.CharField(max_length=255)
+  contract_address = models.CharField(max_length=255)
+  amount = models.IntegerField(default=1)
+  expires_date = models.DateTimeField()
```
Comment (mbeacom) [inline app/dashboard/models.py]: Per our slack discussion, do you think you can either override `save()` on this model and `self.contract_address = web3.toChecksumAddress(self.contract_address)` or perform the `toChecksumAddress` in the management command to ensure the contract_address is always stored such that it'll succeed during redemption?

File: app/dashboard/models.py [hunk 6]
```diff
@@ -764,3 +764,21 @@ def normalize_tip_usernames(sender, instance, **kwargs):
...
+
+
+class CoinRedemptionRequest(SuperModel):
+  coin_redemption = models.OneToOneField(CoinRedemption, blank=False, on_delete=models.CASCADE)
+  ip = models.CharField(max_length=50)
```
Comment (mbeacom) [inline app/dashboard/models.py]: Just a note, there is an IP Address model field type. https://docs.djangoproject.com/en/2.0/ref/models/fields/#genericipaddressfield

File: app/dashboard/models.py (hunk 5 above)
Comment (mbeacom) [reply app/dashboard/models.py]: Disregard this, I'll update it after this is merged.

**Comment by eswarasai** [comment]
@mbeacom -- I've updated the code as per your comments. Let me know if there's anything else pending. Thanks.

**Comment by owocki** [comment]
@mbeacom last we spoke you were having trouble redeeming the coins.. is that still the case?

File: app/app/local_settings.py.dist [hunk 7]
```diff
@@ -135,4 +135,11 @@ AWS_SECRET_ACCESS_KEY = 'TODO'
 S3_REPORT_BUCKET = 'TODO'
 S3_REPORT_PREFIX = 'TODO'
 
+# WEB3
+WEB3_HTTP_PROVIDER = 'TODO'
```
Comment (owocki) [inline app/app/local_settings.py.dist]: could this just default to mainnet.infura.io?

File: app/app/local_settings.py.dist [hunk 8]
```diff
@@ -135,4 +135,11 @@ AWS_SECRET_ACCESS_KEY = 'TODO'
...
+
+# COLO Coin
+COLO_ACCOUNT_ADDRESS = 'TODO'
```
Comment (owocki) [inline app/app/local_settings.py.dist]: @mbeacom could you PR a related change to the server repo with a new account for this?  from there i can fund it with some tokens / eth

File: app/dashboard/management/commands/create_coin_redemption_shortcodes.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                shortcode=shortcode,
+                url='https://gitcoin.co/coin/redeem/' + shortcode,
+                network=options['network'],
+                token_name=options['token_name'],
+                contract_address=options['contract_address'],
+                amount=options['amount'],
+                expires_date=timezone.now() + timezone.timedelta(seconds=options['expires_date'])
+            )
+
+            print("https://gitcoin.co/coin/redeem/" + cr.shortcode)
```
Comment (owocki) [inline app/dashboard/management/commands/create_coin_redemption_shortcodes.py]: awesome!

**Comment by mbeacom** [comment]
@owocki Testing now with `COLO` on Ropsten.

File: app/dashboard/views.py [hunk 10]
```diff
@@ -776,3 +785,75 @@ def toolbox(request):
...
+def redeem_coin(request, shortcode):
+    if request.body:
+        status = 'OK'
+
+        body_unicode = request.body.decode('utf-8')
+        body = json.loads(body_unicode)
+        address = body['address']
+
+        try:
+            coin = CoinRedemption.objects.get(shortcode=shortcode)
```
Comment (owocki) [inline app/dashboard/views.py]: using this code... whats to stop me from just using the same `shortcode` over and over again

Comment (mbeacom) [reply app/dashboard/views.py]: ^ Agreed.  Maybe we should check for the existence  of a `CoinRedemptionRequest` for the `coin` before performing the transfer.

Comment (eswarasai) [reply app/dashboard/views.py]: ^^Bummer. I've added that check for `GET` request and totally forgot about the similar issue for the API apart. Let me quickly do that. Thanks for pointing this out 👍

File: app/app/local_settings.py.dist (hunk 7 above)
Comment (eswarasai) [reply app/app/local_settings.py.dist]: @owocki -- Sure. I can go ahead and do that in the following commit

File: app/dashboard/views.py (hunk 10 above)
Comment (eswarasai) [reply app/dashboard/views.py]: @owocki @mbeacom -- This has been taken care of now

File: app/app/local_settings.py.dist (hunk 7 above)
Comment (owocki) [reply app/app/local_settings.py.dist]: not as of now.  in the future, we might need a key

Comment (eswarasai) [reply app/app/local_settings.py.dist]: @owocki -- I've updated this with the `https://mainnet.infura.io` now

**Comment by mbeacom** [comment]
@eswarasai Hate to ask you, but would you mind resolving the conflicts on this PR?

**Comment by eswarasai** [comment]
@mbeacom -- Done 👍

**Comment by owocki** [comment]
yayayaaya!  lets ship this before the craziness of ETHDenver starts :)

**Comment by owocki** [comment]
🎉

**Pull request merged**